File: mobile/native/apple/apps/ensu/Ensu/Services/KnowledgeStore.swift
import Foundation
import SwiftUI

@MainActor
final class KnowledgeStore: ObservableObject {
    @Published private(set) var packs: [KnowledgePackState]
    @Published var downloadsAllowed = false

    private let provider: KnowledgeProvider
    private let preferences: KnowledgePreferences
    private var bootstrapTask: Task<Void, Never>?
    private var mutationTasks: [String: Task<Void, Never>] = [:]

    init(
        datasets: [KnowledgeDatasetConfig],
        provider: KnowledgeProvider,
        defaults: UserDefaults = .standard
    ) {
        packs = datasets.map { KnowledgePackState(config: $0) }
        self.provider = provider
        preferences = KnowledgePreferences(defaults: defaults)
    }

    var enabledReadyDatasets: [KnowledgeDatasetConfig] {
        packs.compactMap { pack in
            guard pack.enabled,
                  pack.status == .ready || pack.status == .updateAvailable else {
                return nil
            }
            return pack.config
        }
    }

    func bootstrap() async {
        if let bootstrapTask {
            await bootstrapTask.value
            return
        }

        let task = Task { @MainActor [weak self] in
            guard let self else { return }
            await bootstrapEnabledDatasets()
        }
        bootstrapTask = task
        await task.value
    }

    private func bootstrapEnabledDatasets() async {
        let requestedEnabled = preferences.enabledDatasetIds
        let datasets = packs.map(\.config)
        for dataset in datasets where requestedEnabled.contains(dataset.stableId) {
            await reconcileForBootstrap(dataset, enabled: true)
        }

        Task { @MainActor [weak self] in
            guard let self else { return }
            for dataset in datasets where !requestedEnabled.contains(dataset.stableId) {
                await reconcileForBootstrap(dataset, enabled: false)
            }
        }
    }

    private func reconcileForBootstrap(
        _ dataset: KnowledgeDatasetConfig,
        enabled: Bool
    ) async {
        do {
            let result = try await provider.reconcile(dataset: dataset)
            applyReconciliation(
                result,
                stableId: dataset.stableId,
                enabled: enabled
            )
        } catch {
            updatePack(dataset.stableId) { pack in
                pack.status = .download
                pack.enabled = false
                pack.errorMessage = userFacingKnowledgeError(error)
            }
        }
    }

    func downloadOrUpdate(stableId: String) {
        guard downloadsAllowed,
              mutationTasks[stableId] == nil,
              let index = packs.firstIndex(where: { $0.id == stableId }) else {
            return
        }
        let dataset = packs[index].config
        let wasInstalled = packs[index].activeIdentity != nil
        updatePack(stableId) { pack in
            pack.isMutating = true
            pack.progressPercent = 0
            pack.progressLabel = "Starting download..."
            pack.errorMessage = nil
        }

        mutationTasks[stableId] = Task { [weak self] in
            guard let self else { return }
            do {
                let result = try await provider.download(dataset: dataset) { progress in
                    Task { @MainActor [weak self] in
                        self?.updatePack(stableId) { pack in
                            pack.progressPercent = min(100, max(0, Int(progress.percentage)))
                            pack.progressLabel = progress.label
                        }
                    }
                }
                let shouldEnable = !wasInstalled && result.status == .ready
                if shouldEnable {
                    preferences.setDatasetEnabled(id: stableId, enabled: true)
                }
                let preservedEnablement = packs
                    .first(where: { $0.id == stableId })?.enabled == true
                applyReconciliation(
                    result,
                    stableId: stableId,
                    enabled: wasInstalled ? preservedEnablement : shouldEnable
                )
            } catch {
                if let result = try? await provider.reconcile(dataset: dataset) {
                    let enabled = packs
                        .first(where: { $0.id == stableId })?.enabled == true
                    applyReconciliation(
                        result,
                        stableId: stableId,
                        enabled: enabled
                    )
                }
                updatePack(stableId) { pack in
                    pack.errorMessage = userFacingKnowledgeError(error)
                }
            }
            updatePack(stableId) { pack in
                pack.isMutating = false
                pack.progressPercent = nil
                pack.progressLabel = nil
            }
            mutationTasks.removeValue(forKey: stableId)
        }
    }

    func cancel(stableId: String) {
        guard let dataset = packs.first(where: { $0.id == stableId })?.config else {
            return
        }
        let ownerTask = mutationTasks[stableId]
        Task { [weak self] in
            guard let self else { return }
            let result = await provider.cancel(dataset: dataset)
            await ownerTask?.value
            if let result {
                let enabled = packs
                    .first(where: { $0.id == stableId })?.enabled == true
                applyReconciliation(result, stableId: stableId, enabled: enabled)
            }
            updatePack(stableId) { pack in
                pack.isMutating = false
                pack.progressPercent = nil
                pack.progressLabel = nil
            }
        }
    }

    func setEnabled(stableId: String, enabled: Bool) {
        guard let index = packs.firstIndex(where: { $0.id == stableId }),
              packs[index].activeIdentity != nil,
              !packs[index].isMutating else {
            return
        }
        updatePack(stableId) { pack in
            pack.enabled = enabled
        }
        preferences.setDatasetEnabled(id: stableId, enabled: enabled)
    }

    private func applyReconciliation(
        _ result: KnowledgeReconciliation,
        stableId: String,
        enabled: Bool
    ) {
        updatePack(stableId) { pack in
            pack.status = result.status
            pack.activeIdentity = result.activeIdentity
            pack.enabled = enabled && result.activeIdentity != nil
            pack.errorMessage = nil
        }
    }

    private func updatePack(
        _ stableId: String,
        update: (inout KnowledgePackState) -> Void
    ) {
        guard let index = packs.firstIndex(where: { $0.id == stableId }) else {
            return
        }
        var updated = packs
        update(&updated[index])
        packs = updated
    }

    private func userFacingKnowledgeError(_ error: Error) -> String {
        guard case let KnowledgeDownloadError.Download(downloadError) = error else {
            return "Knowledge pack setup failed. Please try again."
        }
        switch downloadError {
        case .cancelled:
            return "Download cancelled"
        case .storageFull:
            return "Not enough storage space to download this knowledge pack."
        case .network:
            return "Couldn't download the knowledge pack. Check your connection and try again."
        case .http:
            return "The knowledge pack is currently unavailable. Please try again later."
        case .validation, .sizeMismatch, .`protocol`, .invalidTarget:
            return "The knowledge pack couldn't be verified. Please try again."
        case .io:
            return "Couldn't save the knowledge pack. Please try again."
        }
    }
}

struct KnowledgePackState: Identifiable, Sendable {
    var id: String { config.stableId }
    let config: KnowledgeDatasetConfig
    var status: KnowledgeReconciliationStatus?
    var activeIdentity: String?
    var enabled = false
    var isMutating = false
    var progressPercent: Int?
    var progressLabel: String?
    var errorMessage: String?
}

private final class KnowledgePreferences {
    private enum Keys {
        static let enabledDatasetIds = "ensu.knowledge.enabled_dataset_ids"
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults) {
        self.defaults = defaults
    }

    var enabledDatasetIds: Set<String> {
        Set(defaults.stringArray(forKey: Keys.enabledDatasetIds) ?? [])
    }

    func setDatasetEnabled(id: String, enabled: Bool) {
        var ids = enabledDatasetIds
        if enabled {
            ids.insert(id)
        } else {
            ids.remove(id)
        }
        defaults.set(ids.sorted(), forKey: Keys.enabledDatasetIds)
    }
}

File: mobile/native/android/apps/ensu/app/src/main/java/io/ente/ensu/knowledge/KnowledgeStore.kt
package io.ente.ensu.knowledge

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import io.ente.ensu.AppState
import io.ente.ensu.bindings.DownloadError
import io.ente.ensu.bindings.KnowledgeDatasetConfig
import io.ente.ensu.bindings.KnowledgeDownloadException
import io.ente.ensu.bindings.KnowledgeDownloadProgress
import io.ente.ensu.bindings.KnowledgeReconciliation
import io.ente.ensu.bindings.KnowledgeReconciliationStatus
import io.ente.ensu.device.isChatSupported
import io.ente.ensu.logging.FileLogRepository
import io.ente.ensu.logging.LogLevel
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private val Context.knowledgePreferences by preferencesDataStore("ensu_knowledge_preferences")

data class KnowledgePackState(
    val config: KnowledgeDatasetConfig,
    val status: KnowledgeReconciliationStatus? = null,
    val activeIdentity: String? = null,
    val enabled: Boolean = false,
    val mutationProgress: KnowledgeDownloadProgress? = null,
    val errorMessage: String? = null
) {
    val isMutating: Boolean get() = mutationProgress != null
}

data class KnowledgeState(
    val packs: Map<String, KnowledgePackState> = emptyMap()
) {
    val enabledReadyDatasets: List<KnowledgeDatasetConfig>
        get() = packs.values
            .filter { pack ->
                pack.enabled &&
                    (pack.status == KnowledgeReconciliationStatus.READY ||
                        pack.status == KnowledgeReconciliationStatus.UPDATE_AVAILABLE)
            }
            .map { it.config }
}

class KnowledgeStore(
    context: Context,
    private val state: MutableStateFlow<AppState>,
    private val provider: KnowledgeProvider,
    datasets: List<KnowledgeDatasetConfig>,
    private val logRepository: FileLogRepository
) {
    private val preferences = KnowledgePreferences(context)
    private val catalog = datasets.associateBy { it.stableId }
    private val jobs = mutableMapOf<String, Job>()
    private var scope: CoroutineScope? = null
    private var bootstrapJob: Job? = null
    private val enabledPacksReady = CompletableDeferred<Unit>()

    fun bootstrap(scope: CoroutineScope) {
        this.scope = scope
        if (bootstrapJob != null) return
        state.update { appState ->
            appState.copy(
                knowledge = KnowledgeState(
                    packs = catalog.mapValues { (_, config) -> KnowledgePackState(config) }
                )
            )
        }
        bootstrapJob = scope.launch {
            val requestedEnabled = runCatching {
                preferences.enabledDatasetIds.first()
            }.getOrDefault(emptySet())
            val (enabled, disabled) = catalog.values.partition {
                it.stableId in requestedEnabled
            }
            try {
                enabled.forEach { dataset ->
                    reconcileAndUpdate(dataset, enabled = true)
                }
            } finally {
                enabledPacksReady.complete(Unit)
            }
            disabled.forEach { dataset ->
                reconcileAndUpdate(dataset, enabled = false)
            }
        }
    }

    suspend fun awaitEnabledPacksReady() {
        enabledPacksReady.await()
    }

    fun downloadOrUpdate(stableId: String) {
        if (!state.value.chat.deviceCapability.isChatSupported()) return
        val dataset = catalog[stableId] ?: return
        val ownerScope = scope ?: return
        if (jobs[stableId]?.isActive == true) return
        val wasInstalled = state.value.knowledge.packs[stableId]?.activeIdentity != null
        updatePack(stableId) {
            it.copy(
                mutationProgress = KnowledgeDownloadProgress("Starting download...", 0.0),
                errorMessage = null
            )
        }
        jobs[stableId] = ownerScope.launch {
            try {
                val reconciliation = provider.download(dataset) { progress ->
                    updatePack(stableId) { current ->
                        current.copy(mutationProgress = progress)
                    }
                }
                val shouldEnable = !wasInstalled &&
                    reconciliation.status == KnowledgeReconciliationStatus.READY
                if (shouldEnable) {
                    preferences.setDatasetEnabled(stableId, true)
                }
                val enabled = if (wasInstalled) {
                    state.value.knowledge.packs[stableId]?.enabled == true
                } else {
                    shouldEnable
                }
                updatePack(stableId) {
                    it.fromReconciliation(reconciliation, enabled).copy(
                        mutationProgress = null
                    )
                }
            } catch (error: Throwable) {
                val reconciled = runCatching { provider.reconcile(dataset) }.getOrNull()
                updatePack(stableId) { current ->
                    (reconciled?.let {
                        current.fromReconciliation(it, current.enabled)
                    } ?: current).copy(
                        mutationProgress = null,
                        errorMessage = userFacingKnowledgeError(error)
                    )
                }
                logRepository.log(
                    LogLevel.Error,
                    "Knowledge pack setup failed",
                    details = "pack=$stableId error=${error.message}",
                    tag = "Knowledge"
                )
            } finally {
                jobs.remove(stableId)
            }
        }
    }

    fun cancel(stableId: String) {
        val dataset = catalog[stableId] ?: return
        val ownerScope = scope ?: return
        val ownerJob = jobs[stableId]
        ownerScope.launch {
            val result = runCatching { provider.cancel(dataset) }.getOrNull()
            ownerJob?.join()
            updatePack(stableId) { current ->
                (result?.let { current.fromReconciliation(it, current.enabled) } ?: current).copy(
                    mutationProgress = null
                )
            }
        }
    }

    fun setEnabled(stableId: String, enabled: Boolean) {
        val pack = state.value.knowledge.packs[stableId] ?: return
        if (pack.activeIdentity == null || pack.isMutating) return
        val ownerScope = scope ?: return
        updatePack(stableId) { it.copy(enabled = enabled) }
        ownerScope.launch {
            preferences.setDatasetEnabled(stableId, enabled)
        }
    }

    private fun updatePack(
        stableId: String,
        transform: (KnowledgePackState) -> KnowledgePackState
    ) {
        state.update { appState ->
            val current = appState.knowledge.packs[stableId] ?: return@update appState
            appState.copy(
                knowledge = appState.knowledge.copy(
                    packs = appState.knowledge.packs + (stableId to transform(current))
                )
            )
        }
    }

    private suspend fun reconcileAndUpdate(
        dataset: KnowledgeDatasetConfig,
        enabled: Boolean
    ) {
        val result = runCatching { provider.reconcile(dataset) }
        updatePack(dataset.stableId) { current ->
            result.fold(
                onSuccess = { reconciliation ->
                    current.fromReconciliation(reconciliation, enabled)
                },
                onFailure = { error ->
                    current.copy(
                        status = KnowledgeReconciliationStatus.DOWNLOAD,
                        enabled = false,
                        errorMessage = userFacingKnowledgeError(error)
                    )
                }
            )
        }
    }

    private fun KnowledgePackState.fromReconciliation(
        result: KnowledgeReconciliation,
        enabled: Boolean
    ): KnowledgePackState = copy(
        status = result.status,
        activeIdentity = result.activeIdentity,
        enabled = enabled && result.activeIdentity != null,
        errorMessage = null
    )

    private fun userFacingKnowledgeError(error: Throwable): String {
        val downloadError = (error as? KnowledgeDownloadException.Download)?.error
        return when (downloadError) {
            DownloadError.Cancelled -> "Download cancelled"
            DownloadError.StorageFull ->
                "Not enough storage space to download this knowledge pack."
            is DownloadError.Network ->
                "Couldn't download the knowledge pack. Check your connection and try again."
            is DownloadError.Http ->
                "The knowledge pack is currently unavailable. Please try again later."
            is DownloadError.Validation,
            is DownloadError.SizeMismatch,
            is DownloadError.Protocol,
            is DownloadError.InvalidTarget ->
                "The knowledge pack couldn't be verified. Please try again."
            is DownloadError.Io ->
                "Couldn't save the knowledge pack. Please try again."
            null -> "Knowledge pack setup failed. Please try again."
        }
    }
}

private class KnowledgePreferences(context: Context) {
    private val preferences = context.applicationContext.knowledgePreferences

    val enabledDatasetIds: Flow<Set<String>> = preferences.data.map { values ->
        values[Keys.enabledDatasetIds].orEmpty()
    }

    suspend fun setDatasetEnabled(stableId: String, enabled: Boolean) {
        preferences.edit { values ->
            val ids = values[Keys.enabledDatasetIds].orEmpty()
            values[Keys.enabledDatasetIds] = if (enabled) ids + stableId else ids - stableId
        }
    }

    private object Keys {
        val enabledDatasetIds = stringSetPreferencesKey("enabled_dataset_ids")
    }
}
